Sources/NIOSSL/ObjectIdentifier.swift

```diff
@@ -16,35 +16,45 @@
 @_implementationOnly import CNIOBoringSSLShims
 
 /// Object Identifier (OID)
-public struct NIOSSLObjectIdentifier: Hashable {
-    fileprivate final class Storage {
-        
-        /// All operations accessing `reference` need to be implemented while guaranteeing that the enclosing `self` has a reference count of at least one.
-        /// Otherwise `reference` could already be freed. This would result in undefined behaviour as we access a dangling pointer.
-        /// One way we can guarantee this is by implementation all operations as methods/computed properties on `Storage`.
-        /// Swift guarantees that `self` is not deallocated during the execution of a method/computed property on `self`.
-        private let reference: OpaquePointer
-        
-        fileprivate init?(_ string: String) {
-            let result = string.withCString { string in
-                // If no_name (the last parameter of CNIOBoringSSL_OBJ_txt2obj) is 0 then long names and
-                // short names will be interpreted as well as numerical forms.
-                // If no_name is 1 only the numerical form is acceptable.
-                // source: https://www.openssl.org/docs/manmaster/man3/OBJ_txt2obj.html
-                CNIOBoringSSL_OBJ_txt2obj(string, 1)
+public struct NIOSSLObjectIdentifier {
+    enum Storage {
+        final class Deallocator {
+            var reference: OpaquePointer!
+            
+            init(takeOwnershipOf reference: OpaquePointer!) {
+                self.reference = reference
             }
-            guard let reference = result else {
-                return nil
+            
+            deinit {
+                CNIOBoringSSL_ASN1_OBJECT_free(self.reference)
             }
-            self.reference = reference
         }
         
-        fileprivate init(takingOwnershipOf reference: OpaquePointer) {
-            self.reference = reference
+        case owned(Deallocator)
+        case unowned(reference: OpaquePointer!, owner: AnyObject)
+        
+        init(takeOwnershipOf reference: OpaquePointer!) {
+            self = .owned(.init(takeOwnershipOf: reference))
+        }
+        
+        init(takeUnowned reference: OpaquePointer!, owner: AnyObject) {
```

Comment:
I think a better set of labels is:

```suggestion
        init(borrowing reference: OpaquePointer!, owner: AnyObject) {
```